### ISSUE TITLE
Allow granting/revoking privileges on functions/stored procedures in postgres_privileges

### ISSUE THREAD

**Issue by jdelic** [saltstack/salt#36557]
The current implementations of `salt.states.postgres_privileges` and `salt.modules.postgres.privileges_grant` do not allow `function` as a valid object type for grants. From what I can tell from the source code this is just an oversight, the existing code would work just fine if `FUNCTION` was allowed as a type in [salt/modules/postgres.py#L2540](https://github.com/saltstack/salt/blob/76dca64d03722bb4f09aeed755e1795113055eee/salt/modules/postgres.py#L2540).

Currently, Salt can't grant or revoke privileges on stored procedures.

**Comment by rallytime** [comment]
I am not very familiar with this particular set of modules/states, but I think this makes sense to add and should be simple to fix. 

@akissa You've been largely involved in this state/module and those changes were yours. You might be interested in fixing this up.

**Comment by akissa** [comment]
I did try an implement that but there was something i could not overcome, i cannot remember what it was that prevented me from implementing `function` support.

I do not have the time at the moment to look into this, any one else should feel free to pick this up.

**Comment by eradman** [comment]
I started to implement a fix for this as well, and discovered why this is difficult because PostgreSQL supports [function overloading](https://www.postgresql.org/docs/9.1/static/sql-grant.html), which means that the function must be referenced by name and argument types:

```
=> \df round
                          List of functions
   Schema   | Name  | Result data type | Argument data types |  Type
------------+-------+------------------+---------------------+--------
 pg_catalog | round | double precision | double precision    | normal
 pg_catalog | round | numeric          | numeric             | normal
 pg_catalog | round | numeric          | numeric, integer    | normal
(3 rows)
```

While I agree that it would be nice to support this the only privilege that can be granted or revoked on functions [is EXECUTE](https://www.postgresql.org/docs/9.5/static/sql-grant.html)

**Comment by jdelic** [comment]
@eradman 

> I started to implement a fix for this as well, and discovered why this is difficult. PostgreSQL supports function overloading, which means that the function must be referenced by name and argument types

I see how this complicates things a bit if you want to make it generic (Salt would have to loop over the definitions in the target schema and execute a `GRANT` for each overloaded function), but couldn't we just start with the easy implementation of using `object_name` to also specify the argument types?

Basically:
```
localharvest:
  postgres_privileges.present:
    - maintenance_db: localharvest
    - user: _postgresql
    - db_user: postgres
    - object_name: first_friday(month INTEGER)
    - object_type: function
    - privileges:
      - EXECUTE
    - grant_option: False
```

**Comment by eradman** [comment]
@jdelic, your idea is good, I think. The trick would be to identify the full function signature, not just the `proname`. 

    => SELECT p.oid::regprocedure, proacl AS name
    FROM pg_catalog.pg_proc p
    JOIN pg_catalog.pg_namespace n
    ON n.oid = p.pronamespace
    WHERE nspname = 'pg_catalog'
    AND proname = 'round'
    ORDER BY proname, proargtypes;
               oid           |               name
    -------------------------+-----------------------------------
     round(double precision) |
     round(numeric)          | {=X/postgres,postgres=X/postgres}
     round(numeric,integer)  |
    (3 rows)

**Comment by rallytime** [comment]
@jdelic Please see the fix in #36851 that @eradman has submitted.

**Comment by cachedout** [comment]
Closed via #36851
